### PR TITLE
[webidl2] Add typedefs for missing functions

### PR DESCRIPTION
**Pull request by queengooborg**
Hello!  This is my first pull request to DefinitelyTyped -- please let me know how I can improve my pull request!

---

This PR resolves https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/63342 by adding the missing typedefs for the `write()` and `validate()` functions, including the `WriteOptions` and `WebIDLErrorData` interfaces/classes.  In turn, this synchronizes the typedefs with version 24.4.0 of the package.

---

Checklist:
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - write():
    - https://github.com/w3c/webidl2.js/blob/7193ff191558778d3b322fd9ea7aa8ddee24752d/lib/writer.js#L68-L74
    - https://github.com/w3c/webidl2.js#documentation
  - validate():
    - https://github.com/w3c/webidl2.js/blob/7193ff191558778d3b322fd9ea7aa8ddee24752d/lib/validator.js#L88-L94
  - WebIDLErrorData:
    - https://github.com/w3c/webidl2.js/blob/7193ff191558778d3b322fd9ea7aa8ddee24752d/lib/error.js#L33
    - https://github.com/w3c/webidl2.js/blob/7193ff191558778d3b322fd9ea7aa8ddee24752d/lib/error.js#L105-L116
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## webidl2 ([<kbd>unpkg</kbd>](https://unpkg.com/browse/webidl2@latest/))
was missing the following properties:
1. WebIDL2
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d64469db6e916b8bed70fefeda88e3318ab72a22
</p>

**Comment by saschanaz** [comment]
I'll take a look soonish, but I think we really should do https://github.com/w3c/webidl2.js/issues/255 instead. The blocker is that I really don't know how it works for multiple module files, I only have seen single-file cases.

**Comment by queengooborg** [comment]
Adding the typedefs directly to `webidl2` sounds great as well, and I am all for that! 
 In the meantime, though, it would be awesome to get this PR landed as I currently have to cast `webidl2` to an `any` so that I may use the `validate()` function in my TypeScript-based project.  (I'm not at all familiar with the `--definition` argument or otherwise generating the typedefs from the JavaScript / JSDocs, though, so I feel I won't be much help there... 😛)

**Comment by saschanaz** [comment]
Yup, probably that's the safest option for the time being since migrating to the type generation may cause some breakage. 

Just curious, which of your projects need webidl2? Something relevant to MDN?

**Comment by queengooborg** [comment]
> Just curious, which of your projects need webidl2? Something relevant to MDN?

You guessed it!  Specifically, it's the [mdn-bcd-collector](https://mdn-bcd-collector.gooborg.com), which uses `webidl2` to parse `@webref/idl` to generate tests for interfaces and interface properties.

File: types/webidl2/index.d.ts [hunk 1]
```diff
@@ -105,7 +188,7 @@ export interface AbstractTypeDescription extends AbstractBase {
 
 interface AbstractNonUnionTypeDescription extends AbstractTypeDescription {
     /** String indicating the generic type (e.g. "Promise", "sequence"). The empty string otherwise. */
-    generic: IDLTypeDescription["generic"];
+    generic: IDLTypeDescription['generic'];
```
Comment (saschanaz) [inline types/webidl2/index.d.ts:191]: I'm okay with this change, but I'd suggest this kind of change should be in a separate PR at least. This made the PR big enough to scare me and that delayed the review.

**Review by saschanaz — APPROVED** [review]

File: types/webidl2/index.d.ts (hunk 1 above)
Comment (queengooborg) [reply types/webidl2/index.d.ts:191]: Sorry about that, that was Prettier kicking in and formatting the file automatically as my editor is configured to do!

**Comment by queengooborg** [comment]
Ready to merge!
